docs/manual/developer_guide.adoc

```diff
@@ -1925,6 +1925,172 @@ Tests are added using the add_test cmake call. Each test should finish with a 0
+
+===== Writing e2e tests for specific rules
+
+In order to test that a rule is yielding expected results in the e2e tests, one
+must create a file called `e2e.yml` in a `tests/ocp4/` directory which will
+exist in the rule's directory itself.
+
+The format looks as follows:
+
+----
+---
+default_result: [PASS|FAIL]
+result_after_remediation: [PASS|FAIL]
+----
+
+Where:
+
+* `default_result` will look at the result when the first scan is run.
+* `result_after_remediation` will look at the result when the second scan is run. The second scan takes place after remediations are applied.
+
+Let's look at an example:
+
+For the `controller_use_service_account` rule, which exists in the
+`applications/openshift/controller/` directory, the directory tree will contain
+the rule definition and the test file:
+
+----
+.
+├── rule.yml
+└── tests
+    └── ocp4
+        └── e2e.yml
+----
+
+In this case, we just want to verify that the default value returns a passing
+result. So `e2e.yml` has the following content:
+
+----
+---
+default_result: PASS
+----
+
+Let's look at another example:
+
+For the `api_server_encryption_provider_config` we want to apply a remediation
+which cannot be applied via the `compliance-operator`. So we'll need a manual
+remediation for this.
+
+The directory structure looks as follows:
+
+----
+.
+├── rule.yml
+└── tests
+    └── ocp4
+        ├── e2e-remediation.sh
+        └── e2e.yml
+----
+
+Where our test contains information for both the first default result and the
+expected result after the remediation has been applied:
+
+----
+---
+default_result: FAIL
+result_after_remediation: PASS
+----
+
+The remediation expects the name of the remediation script to be
+`e2e-remediation.sh`. The script should:
+
+* Apply the remediation.
+* Verify that the status has converged.
+
+In the aforementioned case, the remediation script is as follows:
+
+----
+#!/bin/bash
+
+oc patch apiservers cluster -p '{"spec":{"encryption":{"type":"aescbc"}}}' --type=merge
+
+while true; do
+    status=$(oc get openshiftapiserver -o=jsonpath='{range .items[0].status.conditions[?(@.type=="Encrypted")]}{.reason}')
+
+    echo "Current Encryption Status:"
+    oc get openshiftapiserver -o=jsonpath='{range .items[0].status.conditions[?(@.type=="Encrypted")]}{.reason}{"\n"}{.message}{"\n"}'
+
+    if [ "$status" == "EncryptionCompleted" ]; then
+        exit 0
+    fi
+
+    sleep 5
+done
+----
+
+Here, we apply the remediation (through the `patch` command) and probe the
+cluster for status. Once the cluster converges, we exit the script with `0`,
+which is a successful status.
+
+The e2e test run will time out at *15 minuntes* if a script doesn't converge.
+
+Note that the scripts will be run in parallel, but the test run will wait for
+all of them to be done.
+
+===== Running the e2e tests on a local cluster
+
+Note that it's possible to run the e2e tests on a cluster of your choice.
```

Comment:
Should we note that there's no cleanup yet?

Reply:
That's a good point. I'll add that here.